test/test_rosdep_lookup.py

```diff
@@ -189,7 +189,7 @@ def test_RosdepDefinition():
     except ResolutionError as e:
         assert e.rosdep_key == 'trusty_only_key'
         assert e.os_name == 'ubuntu'
-        assert e.os_version == '*'
+        assert e.os_version == 'lucid'
```

Comment:
Note that this is a change in the existing behavior, but this pattern doesn't appear anywhere in the rosdep database right now.

Reply:
> this pattern doesn't appear anywhere in the rosdep database right now.

We don't have anything that fits this exactly but I think any situation where we use the combination of an os_version wildcard with specific `os_version: null` keys is meant to be covered by this test even if we don't have anywhere where a whole os_name is also null (as we usually just omit the key).

Am I right in thinking that this behavior change is due to https://github.com/ros-infrastructure/rosdep/pull/838/commits/91cf3a113e491877de5b3744a1102e616f7ba3d6 which preserved the originally requested os_name and os_version?
This definitely seems like an improvement to the error messages for which a case could be made independently from the os_name wildcard support so I see it as an improvement to existing behavior. But I think that the case under test is an

Reply:
> Am I right in thinking that this behavior change is due to `91cf3a1` ...?

Correct.

However, if you follow the code, the `os_version` isn't getting changed until [L138](https://github.com/ros-infrastructure/rosdep/blob/be780c45312b562f461fd7d2f06232e4e66df1aa/src/rosdep2/lookup.py#L137-L138) at which point the only possible `ResolutionError` would be from an explicitly nulled-out rule. Combined, this means that there is only one pattern which will behave differently under 91cf3a1:
```yaml
foo:
  debian:
    '*': null
    sid: [bar]
```

This syntax is somewhat redundant, and causes rosdep's resolution error message to change, but not the exception type. When last I surveyed we had no occurrences of `'*': null` in the database, but it looks like two of them have slipped in. They should probably be removed.

An OS-level `null` will bypass L138 so the error message won't change. Again, though, we shouldn't have any of these in the database ([even though we do](https://github.com/ros/rosdistro/blob/d5bac54d9ca5485dfc5ac568bf3e743fa0e66282/rosdep/base.yaml#L90)).